packages/perspective-jupyterlab/src/ts/renderer.ts

```diff
@@ -84,13 +85,39 @@ export class PerspectiveDocumentWidget extends DocumentWidget<PerspectiveWidget>
 
             if (this._psp.viewer.table === undefined) {
                 // construct new table
-                this._psp.viewer.load(WORKER.table(data));
+                const table = await WORKER.table(data);
+
+                // load data
+                this._psp.viewer.load(table);
+
+                // create a flat view
+                const view = await table.view();
+                view.on_update(() => {
+                    if (this._type === "csv") {
+                        view.to_csv().then((result: string) => {
+                            this.context.model.fromString(result);
+                            this.context.save();
+                        });
+                    } else if (this._type === "arrow") {
+                        view.to_arrow().then((result: ArrayBuffer) => {
+                            const resultAsB64 = btoa(new Uint8Array(result).reduce((acc, i) => (acc += String.fromCharCode.apply(null, [i])), ""));
+                            this.context.model.fromString(resultAsB64);
+                            this.context.save();
+                        });
+                    } else if (this._type === "json") {
+                        view.to_json().then((result: any) => {
+                            this.context.model.fromJSON(result);
+                            this.context.save();
+                        });
```

Comment:
Prefer `async` over mixing and matching `async` and `Promise` APIs, unless `Promise` is explicitly required (in this case make `on_update()` callback `async`).

Reply:
done